fastestimator/pipeline.py

```diff
@@ -13,442 +13,132 @@
-    if isinstance(data, (list, tuple)):
-        return _infer_batch_size(data[0])
-    if isinstance(data, set):
-        return _infer_batch_size(data.pop())
-    if isinstance(data, dict):
-        return _infer_batch_size(data.popitem()[1])
-    raise ValueError("Could not infer batch size from data loader")
-
-
-# noinspection PyPep8Naming
-def Pipeline(train_data: Union[Dataset, DataLoader, Scheduler, tf.data.Dataset, None] = None,
-             eval_data: Union[Dataset, DataLoader, Scheduler, tf.data.Dataset, None] = None,
-             batch_size: Union[int, Scheduler] = 1,
-             ops: Union[None, NumpyOp, Scheduler[NumpyOp], Iterable[Union[NumpyOp, Scheduler[NumpyOp]]]] = None,
-             drop_last: bool = False,
-             num_process: Union[int, None] = None,
-             shuffle_train: bool = True,
-             shuffle_eval: bool = False) -> BasePipeline:
-    data = train_data if train_data is not None else eval_data
-    assert data is not None, "At least one of train_data or eval_data must be provided"
-    if train_data is not None and eval_data is not None:
-        if isinstance(train_data, (DataLoader, tf.data.Dataset)) or isinstance(eval_data,
-                                                                               (DataLoader, tf.data.Dataset)):
-            assert type(train_data) == type(eval_data), "Cannot mix and match Torch Loaders and tf Datasets"
-    if isinstance(data, tf.data.Dataset):
-        pipeline = TensorFlowPipeline({"train": train_data, "eval": eval_data})
-    elif isinstance(data, DataLoader):
-        pipeline = TorchPipeline({"train": train_data, "eval": eval_data})
-    else:
-        pipeline = FEPipeline(train_data,
-                              eval_data,
-                              batch_size,
-                              ops=ops,
-                              drop_last=drop_last,
-                              num_process=num_process,
-                              shuffle_train=shuffle_train,
-                              shuffle_eval=shuffle_eval)
-    return pipeline
+                signature_epoches.update(range(epoch, epoch + least_common_cycle))
+        signature_epoches = set(epoch for epoch in signature_epoches if epoch < total_epoches)
+        return signature_epoches
+
+    def get_all_output_keys(self, mode, total_epoches) -> Set[str]:
+        output_keys = set()
+        for epoch in self.get_signature_epoches(total_epoches):
+            loader = self.get_loader(mode=mode, epoch=epoch)
+            data = next(iter(loader))
+            assert isinstance(data, dict), "please make sure data output format is dictionary"
+            output_keys = output_keys.union(set(data.keys()))
+            if isinstance(loader, DataLoader) and isinstance(loader.dataset, OpDataset):
+                for op in loader.dataset.ops:
+                    output_keys.update(to_list(op.outputs))
+        output_keys -= {None}
+        return output_keys
```

Comment:
IDE thinks that output_keys will be a set[list] here instead of the desired set[str]. need to look into why

Reply:
need to edit to_list in util.py to be:

def to_list(data: Any) -> List[Any]:

Reply:
done